### PR TITLE
DrivAerML: gradient centralization (implicit regularizer)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Gradient centralization (Yong et al. 2020, "Gradient Centralization: A New Optimization Technique for DNNs") subtracts the per-tensor gradient mean before the optimizer step. This acts as an implicit regularizer by constraining the weight space to a manifold where activation means are preserved, and projects gradients to a lower-variance subspace.

For DrivAerML's batch_size=1 training with high gradient variance, GC may smooth the optimization trajectory without the instability introduced by explicit regularizers (WD crashed at ep178, gc=2.0 diverged). Unlike gradient clipping which truncates, GC centers — it preserves gradient direction while reducing inter-sample drift.

**Reference:** https://arxiv.org/abs/2004.01461

## Instructions

Add `--grad-centralization` flag to train.py. When active, apply gradient centralization as a pre-step hook:

```python
def centralize_gradients(optimizer):
    for group in optimizer.param_groups:
        for p in group['params']:
            if p.grad is None:
                continue
            if p.grad.dim() > 1:  # only centralize weight tensors, not biases
                p.grad.data -= p.grad.data.mean(dim=tuple(range(1, p.grad.dim())), keepdim=True)
```

Call `centralize_gradients(optimizer)` after `loss.backward()` and before `optimizer.step()`.

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --grad-centralization \
  --wandb_group dm-grad-centralization
```

## Reporting Requirements
- Best val `surface_rel_l2_pct` + epoch
- Test at best val checkpoint
- W&B run ID

## Baseline

**DrivAerML champion:**
- `val_primary/surface_rel_l2_pct`: **3.997%** at epoch 467
- W&B: `bht6h42t` (4L/512d, AdamW lr=5e-4, T_max=30, no gc/WD/EMA)
- External: Transolver-3 3.71% | AB-UPT 3.82%

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Gradient Centralization — DrivAerML

| Metric | GC (this run) | Baseline (no GC) | Delta |
|--------|--------------|-------------------|-------|
| **val_primary/surface_rel_l2_pct** | **5.492%** (ep 181) | **3.997%** (ep 467) | +37.4% worse |
| test_primary/surface_rel_l2_pct | 6.212% (ep 181) | — | — |
| External ref | Transolver-3: 3.71% / AB-UPT: 3.82% | | |

**W&B run:** `yblf2j7a`

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --grad-centralization \
  --wandb-group dm-grad-centralization \
  --agent edward --wandb-name "edward/dm-grad-centralization"
```

### What happened

**Gradient centralization did NOT help — negative result.** The run diverged catastrophically after epoch ~200.

**Training trajectory:**
- Epochs 1–181: Healthy convergence, reached best val 5.492% at epoch 181 (grad_norm ~0.21)
- Epoch ~200: Instability onset — val jumped to 8.75%, grad_norm spiked to 1.78
- Epoch ~250: Full divergence — grad_norm exploded to **351,932**, val degraded to 48.6%
- Epochs 250–527: Oscillating between partial recovery and further explosions (grad_norm 234k at ep500). Never recovered below 28%.

**Why it failed:** GC subtracts the per-tensor gradient mean, which preserves gradient direction but removes the DC component. For the cosine annealing schedule (T_max=30), this creates a fatal interaction: when LR restarts at the top of a cosine cycle (~ep 181 ≈ 6×T_max), the centered gradients lose the stabilizing effect of the mean, and the full-LR restart kicks the model out of its basin. The baseline without GC handles these restarts gracefully because the full gradient (with mean) provides a more conservative update direction. The gradient norm explosion (0.21 → 351,932) at the restart boundary confirms this mechanism.

**Key observation:** GC's best (5.492%) is 37% worse than baseline (3.997%) even before divergence. The baseline reaches a deeper basin by training stably through many more cosine cycles. GC prevents this by making the model fragile at cycle boundaries.

### Suggested follow-ups
- GC might work with a monotonically decaying LR schedule (no restarts), but this conflicts with the established benefit of cosine annealing on DrivAerML
- GC + gradient clipping (e.g. gc=1.0) could prevent the norm explosions, but the pre-divergence best was already worse than baseline, suggesting limited upside
- The result is consistent with prior DrivAerML findings that explicit regularizers (WD, gc=2.0) tend to hurt or destabilize training on this dataset

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3112 because ADVISOR: Gradient centralization reaches 5.492% (ep181) then diverges catastrophically at ep200 when cosine restart hits (grad norms 0.21→351,932). Even the pre-divergence best is 43% worse than 3.833%. GC removes the gradient DC component, making the model fragile at LR restart boundaries. Consistent with the pattern: explicit gradient modifications consistently hurt DM training. Closing..
